clang/test/SemaSYCL/decomposition.cpp

```diff
@@ -1,124 +1,156 @@
-// RUN: %clang_cc1 -fsycl -fsycl-is-device -ast-dump %s | FileCheck %s
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -ast-dump -sycl-std=2020 %s | FileCheck %s
+
+// This test checks that the compiler decomposes structs containing special types only
+// (i.e. accessor/stream/sampler etc) and all others are passed without decomposition
+// thus optimizing the number of kernel arguments.
 
 #include "Inputs/sycl.hpp"
 
-using namespace cl::sycl;
+sycl::queue myQueue;
 
-struct has_acc {
-  accessor<char, 1, access::mode::read> acc;
+struct StructWithAccessor {
+  sycl::accessor<char, 1, sycl::access::mode::read> acc;
 };
 
-struct acc_base : accessor<char, 1, access::mode::read> {
+struct StructInheritedAccessor : sycl::accessor<char, 1, sycl::access::mode::read> {
   int i;
 };
 
-struct has_sampler {
-  sampler sampl;
+struct StructWithSampler {
+  sycl::sampler sampl;
 };
 
-struct has_spec_const {
-  ONEAPI::experimental::spec_constant<int, class f1> SC;
+struct StructWithSpecConst {
+  sycl::ONEAPI::experimental::spec_constant<int, class f1> SC;
 };
 
-handler H;
+sycl::handler H;
 
-struct has_stream {
-  stream s1{0, 0, H};
+struct StructWithStream {
+  sycl::stream s1{0, 0, H};
 };
 
-struct has_half {
-  half h;
+struct StructWithHalf {
+  sycl::half h;
 };
 
-struct non_decomposed {
+struct StructNonDecomposed {
   int i;
   float f;
   double d;
 };
 
-struct use_non_decomposed : non_decomposed {
-  non_decomposed member;
+struct StructWithNonDecomposedStruct : StructNonDecomposed {
+  StructNonDecomposed member;
   float f;
   double d;
 };
 
 template <typename T>
-struct Test1 {
+struct StructWithArray {
   T a;
   T b[2];
-  non_decomposed d;
+  StructNonDecomposed SimpleStruct;
   int i;
 };
 
 template <typename T>
-struct Test2 : T {
-  non_decomposed d;
+struct DerivedStruct : T {
+  StructNonDecomposed SimpleStruct;
   int i;
 };
 
-template <typename Name, typename Func>
-__attribute__((sycl_kernel)) void kernel(const Func &kernelFunc) {
-  kernelFunc();
-}
-
 int main() {
 
-  non_decomposed d;
-  non_decomposed ds[5];
-  use_non_decomposed d2;
-  use_non_decomposed d2s[5];
+  StructNonDecomposed SimpleStruct;
```

Comment:
If you're doing follow-up changes, I would suggest just leaving this as d and ds unless you plan to change d2 and d2s as well.